crates/zed/src/languages/astro.rs

```diff
@@ -0,0 +1,136 @@
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use futures::StreamExt;
+use language::{LanguageServerName, LspAdapter, LspAdapterDelegate};
+use lsp::LanguageServerBinary;
+use node_runtime::NodeRuntime;
+use serde_json::json;
+use smol::fs;
+use std::{
+    any::Any,
+    ffi::OsString,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+use util::ResultExt;
+
+const SERVER_PATH: &'static str = "node_modules/@astrojs/language-server/bin/nodeServer.js";
+
+fn server_binary_arguments(server_path: &Path) -> Vec<OsString> {
+    vec![server_path.into(), "--stdio".into()]
+}
+
+pub struct AstroLspAdapter {
+    node: Arc<dyn NodeRuntime>,
+}
+
+impl AstroLspAdapter {
+    pub fn new(node: Arc<dyn NodeRuntime>) -> Self {
+        AstroLspAdapter { node }
+    }
+}
+
+#[async_trait]
+impl LspAdapter for AstroLspAdapter {
+    fn name(&self) -> LanguageServerName {
+        LanguageServerName("astro-language-server".into())
+    }
+
+    fn short_name(&self) -> &'static str {
+        "astro"
+    }
+
+    async fn fetch_latest_server_version(
+        &self,
+        _: &dyn LspAdapterDelegate,
+    ) -> Result<Box<dyn 'static + Any + Send>> {
+        Ok(Box::new(
+            self.node
+                .npm_package_latest_version("@astrojs/language-server")
+                .await?,
+        ) as Box<_>)
+    }
+
+    async fn fetch_server_binary(
+        &self,
+        version: Box<dyn 'static + Send + Any>,
+        container_dir: PathBuf,
+        _: &dyn LspAdapterDelegate,
+    ) -> Result<LanguageServerBinary> {
+        let version = version.downcast::<String>().unwrap();
+        let server_path = container_dir.join(SERVER_PATH);
+
+        if fs::metadata(&server_path).await.is_err() {
+            self.node
+                .npm_install_packages(
+                    &container_dir,
+                    &[("@astrojs/language-server", version.as_str())],
+                )
+                .await?;
+        }
+
+        Ok(LanguageServerBinary {
+            path: self.node.binary_path().await?,
+            arguments: server_binary_arguments(&server_path),
+        })
+    }
+
+    async fn cached_server_binary(
+        &self,
+        container_dir: PathBuf,
+        _: &dyn LspAdapterDelegate,
+    ) -> Option<LanguageServerBinary> {
+        get_cached_server_binary(container_dir, &*self.node).await
+    }
+
+    async fn installation_test_binary(
+        &self,
+        container_dir: PathBuf,
+    ) -> Option<LanguageServerBinary> {
+        get_cached_server_binary(container_dir, &*self.node).await
+    }
+
+    fn initialization_options(&self) -> Option<serde_json::Value> {
+        Some(json!({
+            "provideFormatter": true,
+            "typescript": {
+                "tsdk": "node_modules/typescript/lib",
+            }
```

Comment:
@Princesseuh this part now works 😄

Reply:
What is this path relative to? Does Zed provide this in the same `node_modules` that the language server is in automatically?

Reply:
That's correct.

![Screenshot 2024-01-29T15 27 23](https://github.com/zed-industries/zed/assets/13088001/39627d7f-007a-40d6-973f-a3f4e7fff380)

The path is `~/Library/Application\ Support/Zed/languages/`